p2p/src/main/kotlin/maru/p2p/MaruPeerManagerService.kt

```diff
@@ -0,0 +1,133 @@
+ *
+ * This file is dual-licensed under either the MIT license or Apache License 2.0.
+ * See the LICENSE-MIT and LICENSE-APACHE files in the repository root for details.
+ *
+ * SPDX-License-Identifier: MIT OR Apache-2.0
+ */
+package maru.p2p
+
+import java.time.Duration
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import maru.config.P2P
+import maru.p2p.discovery.MaruDiscoveryPeer
+import maru.p2p.discovery.MaruDiscoveryService
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import org.apache.tuweni.bytes.Bytes
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+import tech.pegasys.teku.networking.p2p.network.P2PNetwork
+import tech.pegasys.teku.networking.p2p.network.PeerHandler
+import tech.pegasys.teku.networking.p2p.peer.DisconnectReason
+import tech.pegasys.teku.networking.p2p.peer.Peer
+
+class MaruPeerManagerService(
+  p2pConfig: P2P,
+) : PeerHandler {
+  private val log: Logger = LogManager.getLogger(this::class.java)
+
+  private val maxPeers = p2pConfig.maxPeers
+  private val currentlySearching = AtomicBoolean(false)
+  private val connectionInProgress = mutableListOf<Bytes>()
+
+  private var discoveryService: MaruDiscoveryService? = null
+  private lateinit var p2pNetwork: P2PNetwork<Peer>
+  private var stopCalled = false
+
+  fun start(
+    discoveryService: MaruDiscoveryService?,
+    p2pNetwork: P2PNetwork<Peer>,
+  ) {
+    this.discoveryService = discoveryService
+    this.p2pNetwork = p2pNetwork
+    searchForPeersUntilMaxReached()
+  }
+
+  fun stop(): SafeFuture<Unit> {
+    stopCalled = true
+    return SafeFuture.completedFuture<Unit>(null)
+  }
+
+  private fun searchForPeersUntilMaxReached() {
+    if (!stopCalled && currentlySearching.compareAndSet(false, true)) {
+      discoveryService?.let { discoveryService ->
+        discoveryService
+          .searchForPeers()
+          .orTimeout(Duration.ofSeconds(30L))
+          .whenComplete { availablePeers, throwable ->
+            log.debug("Finished searching for peers. Found {} peers.", availablePeers.size)
+            if (!stopCalled) {
+              if (throwable != null) {
+                log.debug("Failed to discover peers: {}\n{}", throwable.message, throwable.stackTraceToString())
+                discoveryService.getKnownPeers().forEach { peer ->
+                  tryToConnectIfNotFull(peer)
+                }
+              } else {
+                availablePeers.forEach { peer ->
+                  tryToConnectIfNotFull(peer)
+                }
+              }
+            }
+          }.whenComplete { _, _ ->
+            currentlySearching.set(false)
+            log.debug("Peer count: {}. Max Peers: {}", p2pNetwork.peerCount, maxPeers)
+            if (!stopCalled && p2pNetwork.peerCount < maxPeers) {
+              CompletableFuture.runAsync { searchForPeersUntilMaxReached() }
```

Comment:
Should we delay the next call? How heavy is peer discovery?

Reply:
Also, I'm not entirely confident, but this way of running an infinite loop might have a memory leak. Consider doing it the way it's done in [LineaFinalizationProvider](https://github.com/Consensys/maru/blob/main/linea-finalization-provider/src/main/kotlin/maru/finalization/LineaFinalizationProvider.kt#L26)

Reply:
I thought it should be safe ...
Can you point me to where you are doing something similar in LineaFinalizationProvider? Could not see it :-)
With regards to a delay, I guess it depends on how important it is to find your peers, compared to other stuff that is going on at start-up. Without peers you can't really do anything, but maybe it is OK to start with less than max peers and keep on searching? Or maybe have a min peers? After min peers are found we slow down?